docs/algorithms/ijon/algorithm_en.md

```diff
@@ -0,0 +1,99 @@
+
+## How to use fuzzuf's IJON CLI
+
+To use IJON's fuzzer, first you need to prepare annotated PUTs with instrumentation tools.
+Because fuzzuf doesn't have its own instrumentation tool, please visit [IJON's repo](https://github.com/RUB-SysSec/ijon/) and build the original instrumentation tool.
+
+After you create a PUT and install `fuzzuf`, run
+
+```bash
+fuzzuf ijon -i path/to/initial/seeds/ path/to/PUT @@
+```
+
+to start IJON's fuzzer. The global options available are the same as for AFL.
+For AFL options, see [AFL/algorithm_en.md](/docs/algorithms/afl/algorithm_en.md).
+
+The local option for IJON is:
+
+- `--forksrv 0|1`
+  - If 1 is specified, then fork server mode is enabled. It is enabled by default.
+
+## Example Usage
+
+You can test the original instrumentation tool and IJON's fuzzer in fuzzuf by building and fuzzing [test.c](https://github.com/RUB-SysSec/ijon/blob/master/test.c) and [test2.c](https://github.com/RUB-SysSec/ijon/blob/master/test2.c) found in IJON's repo. Note that, test.c, included in the latest commit (56ebfe34), may yield compilation errors and in that case you need to apply the following changes:
+
+```diff
+diff --git a/llvm_mode/afl-rt.h b/llvm_mode/afl-rt.h
+index 616cbd8..28d5f9d 100644
+--- a/llvm_mode/afl-rt.h
++++ b/llvm_mode/afl-rt.h
+@@ -45,14 +45,14 @@ void ijon_enable_feedback();
+ void ijon_disable_feedback();
+
+ #define _IJON_CONCAT(x, y) x##y
+-#define _IJON_UNIQ_NAME() IJON_CONCAT(temp,__LINE__)
++#define _IJON_UNIQ_NAME IJON_CONCAT(temp,__LINE__)
+ #define _IJON_ABS_DIST(x,y) ((x)<(y) ? (y)-(x) : (x)-(y))
+
+ #define IJON_BITS(x) ((x==0)?{0}:__builtin_clz(x))
+ #define IJON_INC(x) ijon_map_inc(ijon_hashstr(__LINE__,__FILE__)^(x))
+ #define IJON_SET(x) ijon_map_set(ijon_hashstr(__LINE__,__FILE__)^(x))
+
+-#define IJON_CTX(x) ({ uint32_t hash = hashstr(__LINE__,__FILE__); ijon_xor_state(hash); __typeof__(x) IJON_UNIQ_NAME() = (x); ijon_xor_state(hash); IJON_UNIQ_NAME(); })
++#define IJON_CTX(x) ({ uint32_t hash = ijon_hashstr(__LINE__,__FILE__); ijon_xor_state(hash); __typeof__(x) IJON_UNIQ_NAME = (x); ijon_xor_state(hash); IJON_UNIQ_NAME; })
+
+ #define IJON_MAX(x) ijon_max(ijon_hashstr(__LINE__,__FILE__),(x))
+ #define IJON_MIN(x) ijon_max(ijon_hashstr(__LINE__,__FILE__),0xffffffffffffffff-(x))
+diff --git a/test.c b/test.c
+index 50b1b05..aa022f6 100644
+--- a/test.c
++++ b/test.c
+@@ -3,6 +3,7 @@
+ #include<assert.h>
+ #include<stdbool.h>
+ #include <stdlib.h>
++#include <stdint.h>
+
+ #define compare(x,y) IJON_CTX(compare_w((x),(y)))
+ bool compare_w(int x, int y){
+```
+
+For example, you can build test.c and fuzz the produced binary with the following commands:
+
+```bash
+$ (path_to_ijon)/llvm_mode/afl-clang-fast (path_to_ijon)/test.c -o test
+$ mkdir /tmp/ijon_test_indir/ && echo hello > /tmp/ijon_test_indir/hello
+$ fuzzuf ijon -i /tmp/ijon_test_indir/ ./test
```

Comment:
Please describe what is supposed to happen if I run this command.
e.g., "It is successful if fuzzuf finds a bug in about X minutes."
FYI I could find a unique crash in about 3 minutes in my environment.

It's also better to describe the difference between AFL. AFL also found the bug about 30 seconds later IJON found it in my env. If IJON is likely to find the bug faster than AFL, please explain it in the document.

Reply:
Added in https://github.com/fuzzuf/fuzzuf/pull/40/commits/8aa518ce8013750aa68a9a7b6ca383a7384b2787 .
Here, I don't want to compare it with AFL because the difference between them doesn't look significant. In a single run, AFL may find a crash earlier than IJON. Users have to run them many times to be convinced that IJON is faster, which is nevertheless time-consuming. Maybe the difference can be vivid in test2.c, but IJON couldn't find a crash of it within 10 minutes as far as I tested and thus it may be also time-consuming.

Reply:
In that sense, I think test.c and test2.c are not a good exemplification of IJON's superiority to AFL, which they are nice examples of how to use it.